MusicTaste/MusicTaste/APIManagers/SpotifyAPIManager.m

```diff
@@ -6,9 +6,166 @@
+    NSString *spotifyClientID = SpotifyClientID;
+    NSURL *spotifyRedirectURL = [NSURL URLWithString:SpotifyRedirectURLString];
+
+    self.configuration  = [[SPTConfiguration alloc] initWithClientID:spotifyClientID redirectURL:spotifyRedirectURL];
+    
+    self.configuration.playURI = @"";
+
+    self.sessionManager = [[SPTSessionManager alloc] initWithConfiguration:self.configuration delegate:self];
+    
+    SPTScope requestedScope = SPTUserLibraryReadScope | SPTPlaylistReadPrivateScope |  SPTUserTopReadScope;
+     
+    if (@available(iOS 11, *)) {
+        [self.sessionManager initiateSessionWithScope:requestedScope options:SPTDefaultAuthorizationOption];
+    } else {
+        [self.sessionManager initiateSessionWithScope:requestedScope options:SPTDefaultAuthorizationOption presentingViewController:self];
+    }
+    completion(nil, nil);
+}
+
+- (BOOL)application:(UIApplication *)app openURL:(NSURL *)url options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options
+{
+    [self.sessionManager application:app openURL:url options:options];
+    return true;
+}
+
+
+
+#pragma mark - SPTSessionManagerDelegate
+
+- (void)sessionManager:(SPTSessionManager *)manager didInitiateSession:(SPTSession *)session
+{
+    self.token = session.accessToken;
+    [self getSpotifyTracksArtists:^(NSDictionary *dict, NSError *error) {
+    }];
+}
+
+- (void)sessionManager:(SPTSessionManager *)manager didFailWithError:(NSError *)error
+{}
+
+- (void)sessionManager:(SPTSessionManager *)manager didRenewSession:(SPTSession *)session
+{}
+
+
+
+
+
+#pragma mark - Getting Spotify Data
+
+-(void) getSpotifyTracksArtists:(void (^)(NSDictionary *, NSError*))completion{
+    // get artists
+    [self getSpotifyData:@"https://api.spotify.com/v1/me/top/artists" completion:^(NSDictionary * artistDict, NSError * error) {
+        if (!error){
+            NSArray *artistArray = artistDict[@"items"];
+            // get dictionary of genres and artists based on top artists
+            NSDictionary *artistDict = [self convertSpotifyArtists:artistArray];
+            // get tracks data
+            [self getSpotifyData:@"https://api.spotify.com/v1/me/top/tracks" completion:^(NSDictionary * tracksDict, NSError * error) {
+                if (!error){
+                    NSArray *tracksArray = tracksDict[@"items"];
+                    // get dictionary of genres, tracks, and artists based on top tracks
+                    NSDictionary *tracksDict =[self convertSpotifyTracks:tracksArray];
+                    [artistDict[@"artists"] unionSet:tracksDict[@"artists"]];
+                    completion(@{@"artists": artistDict[@"artists"], @"tracks":tracksDict[@"tracks"], @"albums": tracksDict[@"albums"], @"genres": artistDict[@"genres"], @"images": artistDict[@"images"]}, nil);
+                }
+            }];
+        }
+    }];
+}
+
+
+-(void) getSpotifyData:(NSString *)urlString completion:(void (^)(NSDictionary *, NSError*))completion{
```

Comment:
Nice approach to avoid duplicating code for each type of API call. I'm wondering, do you need NSError in the completion callback? Perhaps you could have centralized error handling in this method, and only call the completion if NSError is nil.